infer/environment.cc

```diff
@@ -224,30 +296,70 @@ void KnowledgeRef::removeReferencesToVar(cfg::LocalRef var) {
 
 string TestedKnowledge::toString(const core::GlobalState &gs, const cfg::CFG &cfg) const {
     fmt::memory_buffer buf;
-    if (!truthy->noTypeTests.empty() || !truthy->yesTypeTests.empty()) {
-        fmt::format_to(buf, "  Being truthy entails:\n{}", truthy->toString(gs, cfg));
+    if (!_truthy->noTypeTests.empty() || !_truthy->yesTypeTests.empty()) {
+        fmt::format_to(buf, "  Being truthy entails:\n{}", _truthy->toString(gs, cfg));
     }
-    if (!falsy->noTypeTests.empty() || !falsy->yesTypeTests.empty()) {
-        fmt::format_to(buf, "  Being falsy entails:\n{}", falsy->toString(gs, cfg));
+    if (!_falsy->noTypeTests.empty() || !_falsy->yesTypeTests.empty()) {
+        fmt::format_to(buf, "  Being falsy entails:\n{}", _falsy->toString(gs, cfg));
     }
     return to_string(buf);
 }
 
+void TestedKnowledge::replaceTruthy(cfg::LocalRef of, TypeTestReverseIndex &index, const KnowledgeRef &newTruthy) {
+    // Note: No need to remove old items from Environment::typeTestsWithVar since it is an overapproximation
+    for (auto &entry : newTruthy->yesTypeTests) {
+        index.addToIndex(entry.first, of);
+    }
+
+    for (auto &entry : newTruthy->noTypeTests) {
+        index.addToIndex(entry.first, of);
+    }
+
+    _truthy = newTruthy;
+}
+
+void TestedKnowledge::replaceFalsy(cfg::LocalRef of, TypeTestReverseIndex &index, const KnowledgeRef &newFalsy) {
+    // Note: No need to remove old items from Environment::typeTestsWithVar since it is an overapproximation
+    for (auto &entry : newFalsy->yesTypeTests) {
+        index.addToIndex(entry.first, of);
+    }
+
+    for (auto &entry : newFalsy->noTypeTests) {
+        index.addToIndex(entry.first, of);
+    }
+
+    _falsy = newFalsy;
+}
+
+void TestedKnowledge::replace(cfg::LocalRef of, TypeTestReverseIndex &index, const TestedKnowledge &knowledge) {
```

Comment:
Would `update` make sense as a name here? It seemed like it's not always replacing the entries in the map, but sometimes just updating them.

Reply:
`replace` describes the operation on `TestedKnowledge`. It's _replacing_ the falsy/truthy lists on the current Knowledge object with that of the passed-in knowledge object.

You are right that the _index_ is updated and not replaced.